JDK-8255711: Fix and unify hotspot signal handlers

tstuefe opened the pull request:
Hi all,

may I please have opinions and reviews on this cleanup-fix-patch for the hotspot signal handlers.

Its main intent is to simplify coding and to commonize some of it across all Posix platforms where possible. Also to fix a number of smaller issues.

This will have a number of benefits, mainly easing maintenance pain for porters and reducing bitrot for platform dependent code.

This all builds upon the work @gerard-ziemski did with https://bugs.openjdk.java.net/browse/JDK-8252324.

----

This cleanup was made more complicated by the fact that there exists a non-obvious and undocumented way for a third party app to chain signal handlers (beside the documented one of using libjsig). It seems that the JVM_handle_xxx_functions() are in fact interfaces for third party coding to invoke hotspot signal handling. This only makes sense in combination with `-XX:+AllowUserSignalHandlers`. A cursory github search revealed that this flag is used quite a bit.

See a more in-depth discussion here: [4]. Thanks to @dholmes-ora for untangling this bit of history.

Unfortunately there is no official documentation from Sun or Oracle, and zero regression tests. So I try to preserve this interface as best as I can. I plan to add a proper regression test with a later change, but for now I don't have the time for that.

---

The fixed issues:

1) `PosixSignals::_are_signal_handlers_installed` is used inside the platform handlers to guard a part of the platform handlers against execution in case the signal handlers are not yet installed. 

Initially this confused me since when this handler is called it would of course be installed. So that boolean would always be true. The only explanation I found was that since these handlers can be invoked directly from outside, this is some (ineffective) form of guard against calling this handler too early. 
But that guard can be left out and that boolean removed. Our signal handlers are safe to call before VM initialization is completed.

2) The return code of JVM_handle_xxx_signal() was inconsistently set (some as bool, some as int) as well as unused in normal code paths (excluding outside calls).

3) JVM_handle_xxx_signal are supposed to be exported, but on AIX there is a day-zero bug which caused it to not be exported.

4) Every platform handler has this section:

```
  JavaThread* thread = NULL;
  VMThread* vmthread = NULL;
  if (PosixSignals::are_signal_handlers_installed()) {
    if (t != NULL ){
      if(t->is_Java_thread()) {
        thread = t->as_Java_thread();
      }
      else if(t->is_VM_thread()){
        vmthread = (VMThread *)t;
      }
    }
  }
```

`vmthread` is unused on all platforms and can be removed.

5) Every platform handler has some variant of this section, to ignore SIGPIPE, SIGXFSZ (whose default actions are to terminate the VM):

```
  if (sig == SIGPIPE || sig == SIGXFSZ) {
    // allow chained handler to go first
    if (PosixSignals::chained_handler(sig, info, ucVoid)) {
      return true;
    } else {
      // Ignoring SIGPIPE/SIGXFSZ - see bugs 4229104 or 6499219
      return true;
    }
  }
```

- On s390 and ppc, we miss SIGXFSZ handling 
   _Update: Fixed separately for easier backport, see [https://bugs.openjdk.java.net/browse/JDK-8255734](JDK-8255734)_.
- both paths return true - section can be shortened

Side note: having handlers for those signals may be unnecessary. We could just set the signal handler to `SIG_IGN`. We would have to tiptoe around any third party handlers for those signals, but it still may be simpler.

6) At the end of every platform header, before calling into fatal error handling, we unblock the signal:

```
>  // unmask current signal
>  sigset_t newset;
>  sigemptyset(&newset);
>  sigaddset(&newset, sig);
>  sigprocmask(SIG_UNBLOCK, &newset, NULL);
>
```

- Use of `sigprocmask()` is UB in a multithreaded program. 
- but then, this section is unnecessary anyway, since [https://bugs.openjdk.java.net/browse/JDK-8252533](JDK-8252533) we unmask error signals at the start of the signal handler.

7) the JFR crash protection is not consistently checked in all platform handlers.

8) On Zero, when entering fatal error handling, we do so via fatal() instead of VMError::report_and_die(), thereby discarding the real crash context and obfuscating the register content in the hs-err file (we still see registers, but those stem from the assertion-poison-page mechanism).

9) on Linux ppc64 and AIX, we have this section:

>      if (sig == SIGILL && (pc < (address) 0x200)) {
>        goto report_and_die;
>      }

which is related to the fact that the zero page on AIX is readable, filled with 0, and reading instructions from it will yield us a SIGILL, not a SIGSEGV (0 is not a noop on PPC, so we don't nop-slide).

This coding is irrelevant on Linux. On AIX, it can also be removed, since this SIGILL would be unrecognized by the hotspot and later count as fatal error anyway.

10) When invoking the fatal error handler, we extract the pc from the context and hand it over as "faulting pc". For SIGILL and SIGFPE, this is not totally correct. According to POSIX [3], for those signals the address of the faulting instruction is handed over in `si_info.si_addr`. 

On most platforms this does not matter, they are the same. But on some architectures the pc in the signal context actually points somewhere else, e.g. beyond the faulting instruction. Therefore `si_info.si_addr` is the better choice.

----

The changes in this patch:

a) hotspot signal handling is now done by the following functions:

```
<OS>                 <foreign code, probably signal handler too>
  |                           |
  v                           v
 javaSignalHandler       JVM_handle_linux_signal()
       |                   /
       v                  v
     javaSignalHandler_inner 
               |
               v
     PosixSignals::pd_hotspot_signal_handler()

```

The right branch only exists to support the `AllowUserSignalHandlers` case, see [4].

`javaSignalHandler` is registered as handler, as it was before; JVM_handle_linux_signal() is exported as before.
`javaSignalHandler_inner` contains the shared portion of the signal handler; `PosixSignals::pd_hotspot_signal_handler()` contains the remaining platform dependent portions.


b) I commonized prologue- and epilogue coding. 
  - I simplified (4) to a single line in the shared handler
  - I moved the JFR thread crash protection (7) up to the shared handler
  - I moved the complete epilogue up to the shared handler. That includes calling the chained handlers, should they exist, as well as invoking the fatal error handler. That fixes (8) and (6)
  - Zero has this tradition of showing a robot telling a cat about the error signal, which I like, and kept.
  - I simplified (5) and commonized it, and removed (9) completely
  - In PosixSignals::install_signal_handlers(), I removed the `signal_handlers_are_installed` guard and replaced it with an assert. Unfortunately this causes lots of indentation changes. @gerard-ziemski: if this clashes too much with your patch for JDK-8253742, I'll leave that part out.

Thanks for reviewing.

Testing: this patch ran through our nightlies, in an earlier form. They will be re-ran some more times.

I'd be happy if aarch64 porters could take a look at the aarch64 portion of this change.

Please note that I had to draw a line somewhere - this is an open ended issue and a lot more could be cleaned. See also Gerard's work on [5], which is under review too. 

----

[1] https://mail.openjdk.java.net/pipermail/hotspot-runtime-dev/2020-October/043145.html
[2] https://mail.openjdk.java.net/pipermail/hotspot-runtime-dev/2020-October/043191.html
[3] https://pubs.opengroup.org/onlinepubs/009695399/basedefs/signal.h.html
[4] https://mail.openjdk.java.net/pipermail/jdk-dev/2020-November/004887.html
[5] https://bugs.openjdk.java.net/browse/JDK-8253742

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255711](https://bugs.openjdk.java.net/browse/JDK-8255711): Fix and unify hotspot signal handlers


### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 3a6a80957a0287921e69f45af414651042dcf45c
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to d9a2deffd7462c110ec506543790c3f4fed2d4ce
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1034/head:pull/1034`
`$ git checkout pull/1034`

magicus commented:
/label remove build

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1136,87 +1224,79 @@ void set_signal_handler(int sig, bool set_installed) {
-      // Tell libjsig jvm is setting signal handlers
-      (*begin_signal_setting)();
-    }
+  static bool done = false;
+  assert(!done, "Only call once");
```
Not sure this kind of init guard is really needed, but if it is then the flag should DEBUG_ONLY. Also I don't see you setting this to true anywhere.

tstuefe replied:
Ouch. Stupid rebasing error.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -437,19 +435,25 @@ bool PosixSignals::chained_handler(int sig, siginfo_t* siginfo, void* context) {
-extern "C" JNIEXPORT int JVM_handle_bsd_signal(int signo, siginfo_t* siginfo,
-                                               void* ucontext,
-                                               int abort_if_unrecognized);
+                         JVM_handle_bsd_signal
```
Can we define this using token pasting e.g.

PASTE_TOKENS(JVM_handle, PASTE_TOKENS(INCLUDE_SUFFIX_OS, _signal))

?

tstuefe replied:
I find that actually less readable than what we have now. Plus, right now its easier to grep for the function name.

coleenp commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -502,21 +528,106 @@ void PosixSignals::unblock_error_signals() {
+          "\n#"
+          "\n#    /--------------------\\"
+          "\n#    |      %-7s       |"
+          "\n#    \\---\\ /--------------/"
```
Isn't the little robot supposed to say "segmentation fault" and would that be safe than calling get_signal_name in this context?

coleenp replied:
Thanks for keeping the picture.

tstuefe replied:
Thanks Coleen. Little robot now spells out the name of the signal (since "segmentation fault" is only correct for segv, and I wanted to see it for the other cases too. get_signal_name() is completely harmless, just uses a bit of stack buffer (and that only for the case of unknown signals which are printed numerically; I plan to change that and give us a simple version which just returns static strings).

coleenp replied:
OK.

coleenp reviewed (approved):
Thank you for doing this cleanup and removing some duplicated code in the os_cpu signal handlers.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -502,21 +528,106 @@ void PosixSignals::unblock_error_signals() {
+
+  // Call platform dependent signal handler.
+  if (!signal_was_handled) {
+    signal_was_handled = PosixSignals::pd_hotspot_signal_handler(sig, info, uc, jt);
```
Can we move:

JavaThread* const jt = (t != NULL && t->is_Java_thread()) ? (JavaThread*) t : NULL;

here? It's not used anywhere else.

tstuefe replied:
Sure, I can do this. We can move it up again when/if later needed.

gerard-ziemski commented on src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp:
```diff
@@ -678,29 +636,6 @@ JVM_handle_bsd_signal(int sig,
-    pc = os::Bsd::ucontext_get_pc(uc);
-  }
-
-  // unmask current signal
-  sigset_t newset;
-  sigemptyset(&newset);
-  sigaddset(&newset, sig);
-  sigprocmask(SIG_UNBLOCK, &newset, NULL);
```
This is not being moved to javaSignalHandler_inner(), so why are we dropping this completely?

gerard-ziemski replied:
I just saw the answer to this in the thread, nvm.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -502,21 +528,106 @@ void PosixSignals::unblock_error_signals() {
+      }
+    }
+#if defined(ZERO) && !defined(PRODUCT)
+    char buf[20];
```
How do we know 20 makes it big enough?

tstuefe replied:
Its the signal name, plus possible additions in the form of numbers (e.g. "UNKNOWN (222)") . Now that I write this I see you are right, this could lead to truncation of the number are very long (MAX_INT is 10 digits long). I'll up this to 64.

gerard-ziemski commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -1136,87 +1244,75 @@ void set_signal_handler(int sig, bool set_installed) {
+    (*begin_signal_setting)();
+  }
+
+  set_signal_handler(SIGSEGV, true);
+  set_signal_handler(SIGPIPE, true);
+  set_signal_handler(SIGBUS, true);
+  set_signal_handler(SIGILL, true);
+  set_signal_handler(SIGFPE, true);
+  PPC64_ONLY(set_signal_handler(SIGTRAP, true);)
+  set_signal_handler(SIGXFSZ, true);
```
Can we drop the last argument here, it's always true the way we use set_signal_handler() ?

tstuefe replied:
Yes, we can do this.

tstuefe commented:
Thanks for the review, Gerard! I'll wait for Davids feedback, since he may want more changes done.

gerard-ziemski commented:
Thank you Thomas for the work, I really like the cleanup here!

tstuefe commented:
I added some more minor cleanups:

- added missing include (atomic.hpp)

- CDT complained about check_pending_signals() not returning a value upon exit; we never exit, so thats fine, but I added a return and a ShouldNotReachHere just for clarity.

Review-triggered fixes:

- @gerard-ziemski: removed the "set_installed" parameter from set_signal_handler. That parameter was used to reset the signal 
back to the default handler. But that had not been used. And it had been incorrect too, since the correct way of doing this would have been to reset to any chained handler if one was present when the hotspot signal handler was installed.

- @gerard-ziemski: increased the buffer size for the signal name in little robots text bubble to 64.

- @gerard-ziemski: moved definition of jt closer to where its needed.

- @dholmes-ora: as requested, I merged javaSignalHandler_inner() and JVM_handle_xxx_signal(). I am not super happy about this - one thing I immediately noticed is that now I cannot assert for "+AllowUserSignalHandlers" to make sure this function is only called then. But I can live with it.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -513,13 +466,53 @@ class ErrnoPreserver: public StackObj {
-static bool javaSignalHandler_inner(int sig, siginfo_t* info, void* ucVoid, bool abort_if_unrecognized) {
+// This routine is the shared part of the central hotspot signal handler. It can
+// also be called by a user application, if a user application prefers to do
+// signal handling itself - in that case it needs to pass signals the hotspot
+// internally uses on to the hotspot first.
```
Nit: s/the hotspot/the VM/

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -513,13 +466,53 @@ class ErrnoPreserver: public StackObj {
+// The user-defined signal handler must pass unrecognized signals to this
+// routine, and if it returns true (non-zero), then the signal handler must
+// return immediately.  If the flag "abort_if_unrecognized" is true, then this
+// routine will never retun false (zero), but instead will execute a VM panic
```
typo: retun

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -513,13 +466,53 @@ class ErrnoPreserver: public StackObj {
+// routine, and if it returns true (non-zero), then the signal handler must
+// return immediately.  If the flag "abort_if_unrecognized" is true, then this
+// routine will never retun false (zero), but instead will execute a VM panic
+// routine kill the process.
```
"to kill"

dholmes-ora reviewed (approved):
Thanks for the updates. A couple of minor (mostly pre-existing) nits in the comments.

David

tstuefe commented:
> Thanks for the updates. A couple of minor (mostly pre-existing) nits in the comments.
> 
> David

Thank you David!

tstuefe commented:
/integrate
